Transitions fail to complete when zooming

brunoabinader opened the pull request:
When zooming with a scroll wheel or with pinch gestures, we sometimes fail to finish the fade transition (both fading in and fading out):

![image](https://cloud.githubusercontent.com/assets/52399/7613545/9aa8dc5a-f991-11e4-84d4-6c5cbaddf5a6.png)

brunoabinader commented:
Review @kkaefer @jfirebaugh?

kkaefer commented:
While this fixes the bug, it also means that the map is continuously rendering at 60fps, using system resources and battery. Instead, we need to figure out in what circumstances we need to rerender (= when fading in/out), and not render otherwise.

brunoabinader commented:
Closing this PR in favour of pushing these into a `v8` branch.

brunoabinader commented:
I've cracked my head a bit these past days trying to figure out why this was not working. It turns out that we've missing checking if `Painter::needsAnimation()` along with `Style::hasTransitions()` in `MapContext::renderSync()`. I've learned a lot from the code while resolving this bug, and thus got the chance to do some other fixes and cleanups along the way:

Relevant changes:
- Unify default transition values in `MapData`:
   - Added `defaultFadeDuration` and `defaultTransitionDelay` to `MapData`;
   - `Painter` & `StyleParser` now receives a reference to `MapData`;
   - As previously seen on the code:
      - 300ms is the default fade duration;
      - 0ms is the default transition duration;
   - We no longer pass the current time point to `Style`, since it now uses `MapData.animationTime`, which gets updated in `MapContext::update()`.
   - Updated `StyleParser` check to use a mock `MapData`;
- Pass {`Duration`,`TimePoint`} by const ref whenever possible;
- Unify time point for Mapcontext::update:  
    - Let style classes cascade use the same time point as the one used to
      recalculate style.
    - Cleaned up `MapContext::update()` to return early whenever possible.
    - Cleaned up `MapContext::loadStyleJSON()` to avoid redundant calls to style
      classes cascade.

/cc @jfirebaugh @kkaefer @tmpsantos @tmcw

brunoabinader commented:
> What about moving these commits not exactly related to the fix to a separated PR?

:+1: will do, thanks!

brunoabinader commented:
There are changes in #1888 (specifically, we no longer pass raw fade transition duration values to `FrameHistory::needsAnimation()` in `Painter::needsAnimation()`) that prevents this fix from compiling. Will wait for #1888 to land first before updating this.

brunoabinader commented:
Got this in https://travis-ci.org/mapbox/mapbox-gl-native/jobs/70906410

brunoabinader commented:
Build failure cause is not related to these changes. See #1887 for details.

brunoabinader commented:
This should fix #779 as well.
